### PR TITLE
Add MessageTemplate.clone API

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Just like we did for scheduled jobs. It's pretty useful to be able to copy a messagetemplate (to use as a new template / to create a backup etc).

Before
----------------------------------------
Cannot copy messagetemplate

After
----------------------------------------
Can copy messagetemplate via API

Technical Details
----------------------------------------
This is implemented in exactly the same way as the scheduled job "Job.clone" function.

Comments
----------------------------------------
This is just the API. A follow-up PR could add this as a UI option.

### CODE REVIEW

**Comment by seamuslee001** [comment]
jenkins re test this please

**Comment by seamuslee001** [comment]
@mattwire i like this but two things

1. Should we be permitting cloning of workflow templates at all?
2. I think this needs a test at least (even a very basic one proving the functionality works)

File: CRM/Core/BAO/MessageTemplate.php [hunk 1]
```diff
@@ -586,4 +586,25 @@ public static function sendTemplate($params) {
...
+   * Make a copy.
+   *
+   * @param int $id The entity id to copy.
+   * @param array $params
+   * @return CRM_Core_DAO
+   */
+  public static function copy($id, $params = []) {
+    $fieldsFix = [
+      'suffix' => [
+        'name' => ' - ' . ts('Copy'),
```
Comment (yashodha) [inline CRM/Core/BAO/MessageTemplate.php]: @mattwire I don't think we have name column in message template table - we have field for `msg_title`.

Comment (mattwire) [reply CRM/Core/BAO/MessageTemplate.php]: @yashodha I've fixed the `name` -> `msg_title`. I'm not sure that we need to append a number as well - other entities don't do that and we are not enforcing unique titles (only IDs).

**Comment by jitendrapurohit** [comment]
I've tested this by creating a clone of the message template using API and also created a unit test at https://github.com/civicrm/civicrm-core/pull/16524 to assert the values 👍

**Comment by eileenmcnaughton** [comment]
@mattwire @colemanw @seamuslee001 I generally think the api should support cloning entities. I think we shouldn't be adding something like this to v3 api but for v4 it would be great. My feeling the metadata for this could be defined in the schema in a <clone> or <copy> key (whether to  call the function & api action clone or copy would be up for grabs too)

Here is one way we might mark it up in the metadata

```
--- a/xml/schema/Core/MessageTemplate.xml
+++ b/xml/schema/Core/MessageTemplate.xml
@@ -5,6 +5,7 @@
   <class>MessageTemplate</class>
   <name>civicrm_msg_template</name>
   <comment>Users will need a way to save and retrieve templates with tokens for use in recurring email communication tasks</comment>
+  <is_clonable>true</is_clonable>
   <add>1.6</add>
   <field>
     <name>id</name>
@@ -25,6 +26,9 @@
     <length>255</length>
     <comment>Descriptive title of message</comment>
     <add>1.6</add>
+    <clone>
+      <append>Copy</append>
+    </clone>
   </field>
   <field>
     <name>msg_subject</name>
@@ -73,6 +77,9 @@
     <type>int unsigned</type>
     <comment>a pseudo-FK to civicrm_option_value</comment>
     <add>3.1</add>
+    <clone>
+      <is_clonable>false</is_clonable>
+    </clone>
   </field>
   <field>
     <name>is_default</name>
@@ -81,6 +88,9 @@
     <default>1</default>
     <comment>is this the default message template for the workflow referenced by workflow_id?</comment>
     <add>3.1</add>
+    <clone>
+      <value>0</value>
+    </clone>
   </field>
   <field>
     <name>is_reserved</name>
@@ -88,6 +98,9 @@
     <type>boolean</type>
     <comment>is this the reserved message template which we ship for the workflow referenced by workflow_id?</comment>
     <add>3.1</add>
+    <clone>
+      <value>0</value>
+    </clone>
   </field>
   <field>
     <name>is_sms</name>

```


Although it also feels like designating the 'description field' at the entity level might make sense - ie. if you are referring to Message Templates then the description field is 'msg_title' & knowing that might also help  somewhere else I guess

**Comment by mattwire** [comment]
@eileenmcnaughton All entities that support copy/clone have a copy method on the BAO (or fall-through to copy on the DAO if it is *really* simple).  The API3 wrapper here is really thin and we could probably move:
```
  $params['workflow_id'] = 'null';
  $params['is_reserved'] = 0;
  $params['is_default'] = 0;
```
into the BAO as well.

We don't have a MessageTemplate API in API4 yet and I'm not sure how much work would be required to achieve that? @colemanw 

If we implement for api4 we still need the BAO part of this and from what I've seen running dual tests against api3/api4 is pretty straightforward so we could merge this and add an API4 version later once the infrastructure is ready.

I like the *principle* of adding copy/clone to the metadata but it gets complicated quickly for more complex entities and I'm not sure it would scale (eg. Event: https://github.com/civicrm/civicrm-core/blob/master/CRM/Event/BAO/Event.php#L907-L1006)

**Comment by colemanw** [comment]
So this may sound naive but I would imagine a `clone` api4 action would simply look up an existing record by id and then save a new one by merging in supplied values. This could be a generic action that worked for any entity. E.x.:

```php
Civi\Api4\Contact::clone()
  ->setOriginalId(123)
  ->addValue('first_name', 'New Guy')
  ->execute();
```

**Comment by eileenmcnaughton** [comment]
@colemanw yes - the metadata comes in with regards to what fields can't be copied - e.g in this case workflow_template_id & which fields need some text appended " - Copy"

**Comment by mattwire** [comment]
test this please

**Comment by jaapjansma** [comment]
Betty and I are reviewing PR's and we came across this PR. We have a couple of questions:
1. How dow we test this @mattwire could you give us a step-by-step instruction on how to test this as a user.
2. What is the status of this PR? Does it need work, are all questions answered? Does it need conceptual aproval? @eileenmcnaughton @colemanw what do you think the status is?

**Comment by mattwire** [comment]
@jaapjansma To test as a user go to API Explorer v3. Select MessageTemplate.clone API function and an ID of an existing MessageTemplate (you can get from the MessageTemplates UI). Execute the API and see that there is a copy of the message template in the list of MessageTemplates.

**Comment by jaapjansma** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
I'm not there with adding it as a v3 api - I think v4 now & we should figure out how it should look in v4

**Pull request closed without merging**